File: array/2_Two_Sum.js
// https://leetcode.com/problems/two-sum/

/*
	Notes: using Map is slightly more performant than object,
	albeit at a slight memory cost, possibly due to hashing

*/ 


/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
 // Using Object
const twoSum = (nums, target) => {
    const diffs = {}
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] in diffs)
            return [diffs[nums[i]], i]
        
        diffs[target - nums[i]] = i
    }
}

// Using Map
const twoSum2 = (nums, target) => {
    const diffs = new Map()
    
    for (let i = 0; i < nums.length; i++) {
        if (diffs.has(nums[i]))
            return [diffs.get(nums[i]), i]
        
        diffs.set(target - nums[i], i)
    }
    
}

File: array/2_Two_Sum.java
// https://leetcode.com/problems/two-sum/

class Solution {

    public int[] twoSum(int[] nums, int target) {
        // Map of diff (target - num) and index
        Map<Integer, Integer> diffs = new HashMap<>(nums.length);
        
        for (int i = 0; i < nums.length; i++) {
            if (diffs.containsKey(nums[i]))
                return new int[] { diffs.get(nums[i]), i };
            
            diffs.put(target - nums[i], i);
        }
        
        return null;
    }

}